Allow releasing/reclaiming intrusive_ptr

smessmer opened the pull request:
Summary: This is useful for C APIs where we want to give owning pointers to/from other languages.

Differential Revision: D9121493

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -276,11 +275,37 @@ class intrusive_ptr final {
+   * Returns an owning (!) pointer to the underlying object and makes the
+   * intrusive_ptr instance invalid. That means the refcount is not decreased.
+   * You *must* put the returned pointer back into a intrusive_ptr using
+   * intrusive_ptr::reclaim(ptr) to properly destruct it.
```
It might be helpful to have an API to directly decrement the refcount of a raw pointer to an `intrusive_pointer_target`, instead of making people first make an intrusive pointer, and then move it out of scope.

smessmer replied:
Not sure I like that. Basically, I see release/reclaim as a hack to temporarily pass a pointer through a C API. It should live as a raw pointer as briefly as possible and always go back into an intrusive_ptr because only intrusive_ptr knows how to handle the invariants.

Offering a second way to destruct it adds more code that needs to be aware of the invariants and make sure it knows to destruct at refcount zero and so on.

ezyang replied:
I mean, sure, internally it can be implemented by constructing an intrusive pointer. All I'm saying is that an API of "if you release the pointer, you must call free on it" is simpler than "if you release the pointer, you must rewrap it in intrusive ptr exactly once".

smessmer committed:
Allow releasing/reclaiming intrusive_ptr (#10133)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10133

This is useful for C APIs where we want to give owning pointers to/from other languages.

Reviewed By: ezyang

Differential Revision: D9121493

fbshipit-source-id: b3ab99470e0ac9be24117a9f81f61d325f93965b
